modules/fluid_properties/include/userobjects/Water97FluidProperties.h

```diff
@@ -57,24 +58,72 @@ class Water97FluidProperties : public SinglePhaseFluidProperties
   virtual Real triplePointTemperature() const override;
 
   virtual Real rho_from_p_T(Real pressure, Real temperature) const override;
+  virtual ADReal rho_from_p_T(const ADReal & pressure, const ADReal & temperature) const override;
+
+  virtual Real v_from_p_T(Real pressure, Real temperature) const override;
+  virtual ADReal v_from_p_T(const ADReal & pressure, const ADReal & temperature) const override;
+  virtual void
+  v_from_p_T(Real pressure, Real temperature, Real & v, Real & dv_dp, Real & dv_dT) const override;
+  virtual void v_from_p_T(const ADReal & pressure,
+                          const ADReal & temperature,
+                          ADReal & v,
+                          ADReal & dv_dp,
+                          ADReal & dv_dT) const override;
 
   virtual void rho_from_p_T(
       Real pressure, Real temperature, Real & rho, Real & drho_dp, Real & drho_dT) const override;
+  virtual void rho_from_p_T(const ADReal & pressure,
+                            const ADReal & temperature,
+                            ADReal & rho,
+                            ADReal & drho_dp,
+                            ADReal & drho_dT) const override;
+
+  Real p_from_v_e(Real v, Real e) const override;
+  ADReal p_from_v_e(const ADReal & v, const ADReal & e) const override;
+
+  virtual Real e_from_p_rho(Real p, Real rho) const override;
+  virtual ADReal e_from_p_rho(const ADReal & p, const ADReal & rho) const override;
+  void e_from_p_rho(Real p, Real rho, Real & e, Real & de_dp, Real & de_drho) const override;
+  void e_from_p_rho(const ADReal & p,
+                    const ADReal & rho,
+                    ADReal & e,
+                    ADReal & de_dp,
+                    ADReal & de_drho) const override;
 
   virtual Real e_from_p_T(Real pressure, Real temperature) const override;
+  virtual ADReal e_from_p_T(const ADReal & pressure, const ADReal & temperature) const override;
 
   virtual void
   e_from_p_T(Real pressure, Real temperature, Real & e, Real & de_dp, Real & de_dT) const override;
+  virtual void e_from_p_T(const ADReal & pressure,
```

Comment:
is this really needed?

Reply:
you prefer more stack frames?

Reply:
asking because this is the first case I see of this outside of rho_from_p_T
looks like there's also e_from_p_rho and h_from_p_T I ve never seen in that form as well

Reply:
And these implementations are necessary in order to get derivatives correct for newton inversions

Reply:
You have to be able to compute `ADReal dy_dx` in order get the `ADReal` from newton inversion to be accurate. A `Real dy_dx` won't get it done

Reply:
ah that's the use case